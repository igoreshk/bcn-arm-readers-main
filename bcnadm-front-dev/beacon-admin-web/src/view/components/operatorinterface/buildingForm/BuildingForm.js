import { BUILDING_DIALOG_FORM, LEVEL_FIELD } from 'src/consts/FormsNames';
import React, { Component } from 'react';
import { unregisterField, arrayRemove, change } from 'redux-form';
import PropTypes from 'prop-types';

import BuildingInfoForm from './building/BuildingInfoForm';
import LevelsMaps from './levels/LevelsMaps';
import { getSortedObjectsbyProperty } from './helpers';
import { fieldsAreFilled, validationsArePassed, levelMapsFieldsAreValid } from './ValidateBuildingForm';

const SET_BUILDING_INFO = 'SET_BUILDING_INFO';
const SET_LEVELS_MAPS = 'SET_LEVELS_MAPS';
const steps = [SET_BUILDING_INFO, SET_LEVELS_MAPS];

export class BuildingForm extends Component {
  constructor(props) {
    super(props);

    let currentStep = SET_BUILDING_INFO;
    let currentStepIndex = 0;

    if (this.props.savedCurrentStepIndex) {
      currentStep = this.props.savedCurrentStep;
      currentStepIndex = this.props.savedCurrentStepIndex;
    }

    this.state = {
      currentStep,
      currentStepIndex
    };
  }

  updateLevels = (levels) => {
    const { setCurrentLevels, dispatch } = this.props;
    dispatch(change(BUILDING_DIALOG_FORM, LEVEL_FIELD, levels));
    setCurrentLevels(levels);
  };

  fieldsAreValid() {
    const { width, height, longitude, latitude, workingHours, mapImage, imageLink, name, address } = this.props;
    return (
      fieldsAreFilled(name, address, mapImage || imageLink) &&
      validationsArePassed({ width, height, longitude, latitude, workingHours })
    );
  }

  next = () => {
    const sortedLevels = getSortedObjectsbyProperty(this.props.levels, 'number');
    this.updateLevels(sortedLevels);

    const { currentStepIndex } = this.state;
    const nextStepIndex = currentStepIndex + 1 >= steps.length ? steps.length - 1 : currentStepIndex + 1;
    this.setState((prevState) => ({
      ...prevState,
      currentStepIndex: nextStepIndex,
      currentStep: steps[nextStepIndex]
    }));
  };

  back = () => {
    const { currentStepIndex } = this.state;
    const prevStepIndex = currentStepIndex - 1 < 0 ? 0 : currentStepIndex - 1;
    this.setState((prevState) => ({
      ...prevState,
      currentStepIndex: prevStepIndex,
      currentStep: steps[prevStepIndex]
    }));
  };

  removeLevelField = (fieldIndex) => {
    const { dispatch, levels } = this.props;
    dispatch(arrayRemove(BUILDING_DIALOG_FORM, LEVEL_FIELD, fieldIndex));
    dispatch(unregisterField(BUILDING_DIALOG_FORM, LEVEL_FIELD));
    const changedLevels = levels.filter((level) => level !== levels[fieldIndex]);
    this.updateLevels(changedLevels);
  };

  render() {
    const { imageLink, translate, handleSubmit, closeDialog, isNew, levels } = this.props;
    const { currentStep } = this.state;

    if (currentStep === SET_BUILDING_INFO) {
      return (
        <BuildingInfoForm
          next={this.next}
          translate={translate}
          allowedToProceed={!!this.fieldsAreValid()}
          closeDialog={closeDialog}
          imageLink={imageLink}
          isNew={isNew}
        />
      );
    } else if (currentStep === SET_LEVELS_MAPS) {
      return (
        <LevelsMaps
          allowedToProceed={levelMapsFieldsAreValid(levels)}
          updateLevels={this.updateLevels}
          back={this.back}
          translate={translate}
          levels={levels}
          isNew={isNew}
          handleSubmit={handleSubmit}
          closeDialog={closeDialog}
          removeLevelField={this.removeLevelField}
          addPopupMessageText={this.props.addPopupMessageText}
        />
      );
    }
  }
}

// additional string prop-type for number fields for case of empty field
BuildingForm.propTypes = {
  name: PropTypes.string,
  latitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  longitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  mapImage: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  address: PropTypes.string,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  imageLink: PropTypes.string,
  translate: PropTypes.func.isRequired,
  levels: PropTypes.arrayOf(PropTypes.object),
  setCurrentLevels: PropTypes.func,
  dispatch: PropTypes.func,
  handleSubmit: PropTypes.func,
  closeDialog: PropTypes.func,
  isNew: PropTypes.bool,
  addPopupMessageText: PropTypes.func
};

export default BuildingForm;
