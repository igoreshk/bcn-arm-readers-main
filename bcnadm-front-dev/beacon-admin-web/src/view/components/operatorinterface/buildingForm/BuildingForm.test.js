import React from 'react';
import { shallow } from 'enzyme';
import { BuildingForm } from './BuildingForm';

describe('BuildingForm', () => {
  const props = {
    translate: jest.fn(),
    closeDialog: jest.fn(),
    currentBuilding: {},
    levels: [],
    name: undefined,
    address: undefined,
    width: 0,
    height: 0,
    longitude: undefined,
    latitude: undefined,
    workingHours: undefined,
    mapImage: undefined,
    imageLink: undefined
  };

  const wrapper = shallow(<BuildingForm {...props} />);

  it('check that the necessary steps are rendered when changing buildind info', () => {
    wrapper.setState({ currentStep: 'SET_BUILDING_INFO' });
    expect(wrapper).toMatchSnapshot();
  });

  it('check that the necessary steps are rendered when changing level maps', () => {
    wrapper.setState({
      currentStep: 'SET_LEVELS_MAPS',
      currentStepIndex: 1
    });
    expect(wrapper).toMatchSnapshot();
  });
});
