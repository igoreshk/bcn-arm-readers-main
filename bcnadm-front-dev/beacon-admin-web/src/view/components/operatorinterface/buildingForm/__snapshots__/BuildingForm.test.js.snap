// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildingForm check that the necessary steps are rendered when changing buildind info 1`] = `
<buildingInfoForm
  allowedToProceed={false}
  closeDialog={[MockFunction]}
  next={[Function]}
  translate={[MockFunction]}
/>
`;

exports[`BuildingForm check that the necessary steps are rendered when changing level maps 1`] = `
<LevelsMaps
  allowedToProceed={true}
  back={[Function]}
  closeDialog={[MockFunction]}
  levels={Array []}
  removeLevelField={[Function]}
  translate={[MockFunction]}
  updateLevels={[Function]}
/>
`;
